File: in_Java/0993 Cousins in Binary Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean isCousins(TreeNode root, int x, int y) {
        Queue<TreeNode> q = new LinkedList<>();
        int level_x = 0;
        int level_y = 0;
        int parent_x = 0;
        int parent_y = 0;
        q.add(root);
        int len;
        TreeNode cur;
        int id = 0;
        while(q.size() != 0){
            len = q.size();
            while(len-- != 0){
                cur = q.poll();
                if(cur.left != null){
                    if(cur.left.val == x){
                        level_x = id+1;
                        parent_x = cur.val;
                    }
                    if(cur.left.val == y){
                        level_y = id+1;
                        parent_y = cur.val;
                    }
                    q.add(cur.left);
                }
                if(cur.right != null){
                    if(cur.right.val == x){
                        level_x = id+1;
                        parent_x = cur.val;
                    }
                    if(cur.right.val == y){
                        level_y = id+1;
                        parent_y = cur.val;
                    }
                    q.add(cur.right);
                } 
            }
            id++;
        }
        return level_x == level_y && parent_x != parent_y;
    }
}

File: in_Python/0993 Cousins in Binary Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def isCousins(self, root, x, y):
        """
        :type root: TreeNode
        :type x: int
        :type y: int
        :rtype: bool
        """
        q = [root]
        p1 = None
        p2 = None
        l1 = -1
        l2 = -1
        l = 0
        while q:
            n = len(q)
            l += 1
            while n != 0:
                n -= 1
                h = q.pop(0)
                if h.left:
                    if h.left.val == x:
                        p1 = h
                        l1 = l
                    if h.left.val == y:
                        p2 = h
                        l2 = l
                    q.append(h.left)
                if h.right:
                    if h.right.val == x:
                        p1 = h
                        l1 = l
                    if h.right.val == y:
                        p2 = h
                        l2 = l
                    q.append(h.right)
                if p1 and p2:
                    return p1 != p2 and l1 == l2
        return False